test/mail-test-srv/main.go

```diff
@@ -0,0 +1,225 @@
+
+	readBuf := bufio.NewReader(conn)
+	conn.Write([]byte("220 smtp.example.com ESMTP\r\n"))
+	if err := expectLine(readBuf, "EHLO localhost"); err != nil {
+		log.Printf("mailcatch: %s: %v\n", conn.RemoteAddr(), err)
+		return
+	}
+	conn.Write([]byte("250-PIPELINING\r\n"))
+	conn.Write([]byte("250-AUTH PLAIN LOGIN\r\n"))
+	conn.Write([]byte("250 8BITMIME\r\n"))
+	if err := expectLine(readBuf, authLine); err != nil {
+		log.Printf("mailcatch: %s: %v\n", conn.RemoteAddr(), err)
+		return
+	}
+	conn.Write([]byte("235 2.7.0 Authentication successful\r\n"))
+	auditlogger.Info(fmt.Sprintf("mailcatch: Successful auth from %s", conn.RemoteAddr()))
+	// == END authentication
+
+	// necessary commands:
+	// MAIL RCPT DATA QUIT
+
+	var fromAddr string
+	var toAddr []string
+	var msgBuf bytes.Buffer
+
+	clearState := func() {
+		fromAddr = ""
+		toAddr = nil
+		msgBuf = bytes.Buffer{}
+	}
+
+	var quit = false
+	for !quit {
+		line, err := getLine(readBuf)
+		if err != nil {
+			log.Printf("mailcatch: %s: getline: %v\n", conn.RemoteAddr(), err)
+			return
+		}
+
+		cmdSplit := strings.SplitN(line, " ", 2)
+		cmd := cmdSplit[0]
+		switch cmd {
+		case "QUIT":
+			quit = true
+			conn.Write([]byte("221 Bye \r\n"))
+			break
+		case "RSET":
+			clearState()
+			conn.Write(smtpOk250)
+		case "NOOP":
+			conn.Write(smtpOk250)
+		case "MAIL":
+			clearState()
+			matches := mailFromRegex.FindStringSubmatch(line)
+			if matches == nil {
+				log.Printf("mailcatch: %s: MAIL FROM parse error\n", conn.RemoteAddr())
+				conn.Write(smtpErr501)
+				continue
+			}
+			addr, err := mail.ParseAddress(matches[1])
+			if err != nil {
+				log.Printf("mailcatch: %s: addr parse error: %v\n", conn.RemoteAddr(), err)
+				conn.Write(smtpErr501)
+				continue
+			}
+			fromAddr = addr.Address
+			conn.Write(smtpOk250)
+		case "RCPT":
+			matches := rcptToRegex.FindStringSubmatch(line)
+			if matches == nil {
+				conn.Write(smtpErr501)
+				continue
+			}
+			addr, err := mail.ParseAddress(matches[1])
+			if err != nil {
+				log.Printf("mailcatch: %s: addr parse error: %v\n", conn.RemoteAddr(), err)
+				conn.Write(smtpErr501)
+				continue
+			}
+			toAddr = append(toAddr, addr.Address)
+			conn.Write(smtpOk250)
+		case "DATA":
+			conn.Write([]byte("354 Start mail input \r\n"))
+
+			for {
+				line, err := getLine(readBuf)
+				if err != nil {
+					log.Printf("mailcatch: %s: getline: %v\n", conn.RemoteAddr(), err)
+					return
+				}
+				msgBuf.WriteString(line)
+				msgBuf.WriteString("\r\n")
+				if strings.HasSuffix(msgBuf.String(), "\r\n.\r\n") {
+					break
+				}
+			}
+
+			mailResult := rcvdMail{
+				From: fromAddr,
+				Mail: msgBuf.String(),
+			}
+			allMailMutex.Lock()
+			for _, rcpt := range toAddr {
+				mailResult.To = rcpt
+				allReceivedMail = append(allReceivedMail, mailResult)
+				log.Printf("mailcatch: Got mail: %s -> %s\n", fromAddr, rcpt)
+			}
+			allMailMutex.Unlock()
+			conn.Write([]byte("250 Got mail \r\n"))
+			clearState()
```

Comment:
msgBuf should be scoped to this particular select clause, since it's used nowhere else.